### DISCUSSION TITLE
GitHub Pages does not support routing for single page apps

### DISCUSSION THREAD

**Discussion by derekantrican** [community/community#64096] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

_This is documented in a variety of StackOverflow posts (such as https://stackoverflow.com/q/71984401/2246411) or in other GitHub issues, but the most centralized place (most discussion & subject of most references from other repos) I've found is here: https://github.com/isaacs/github/issues/408_

Simply put, GitHub pages does not support routing for single page apps (and should). The most basic example of this is a react app with a single index.html file, but various pages. Navigating by clicking on links while in the app will work, but going directly to a page's url in the address bar (eg username.github.io/mysite/page1 or mysite.com/page1) will result in a 404 error page as no "page1.html" exists and GitHub pages doesn't have a way for the repo to indicate all requests should go back to the index.html.

As far as I can tell from my testing, there are 2 valid workarounds:

- When using `react-router`, use `HashRouter` instead of `BrowserRouter`. This works, but urls are a little uglier as `/#/page1` rather than `/page1`
- Using a custom 404.html page to parse the requested path, modify the browser history, and redirect to index.html (which will then look at the browser history) as demonstrated here: https://github.com/rafgraph/spa-github-pages. This is VERY hacky and as pages still initially return 404, has implications for SEO & some browsers (some browsers will show a "error banner" when a 404 is returned, even if the page eventually resolves)

Please support simple routing (eg `/page1`) for single page apps in GitHub Pages!

**Comment by christophercouch** [comment]
Hello derekantrican,

I understand the challenge you're facing with routing for single page apps on GitHub Pages. Currently, GitHub Pages doesn't natively support SPA routing. This is indeed a limitation that many developers have raised.

For now, the common workarounds you mentioned are:

1. Using `HashRouter` in react-router to have URLs like `/#/page1`.
2. A custom 404.html page that redirects to index.html, though this can have SEO implications.

I recognize that these workarounds are not ideal, especially for developers looking to host SPAs. It would indeed be beneficial for GitHub Pages to support direct routing for SPAs without the need for such workarounds.

**Reply by ng-druid** [reply]
That has worked out well for us is using azure static web apps instead which supports SPA routing and can be used for free. It is very easy to integrate azure static web apps with a static build repo hosted on Github.

https://azure.microsoft.com/en-us/products/app-service/static

**Reply by olivi-eh** [reply]
Netlify also supports SPA routing: https://docs.netlify.com/routing/redirects/rewrites-proxies/#history-pushstate-and-single-page-apps

It's truly a shame GitHub Pages doesn't. I'm highly considering moving elsewhere since it impacts my SEO (all pages except the homepage are considered "404" from the SEO crawlers' perspective).

**Reply by prahladyeri** [reply]
Bump! It's 2026 already, I really hope GH-Pages will support SPA routing soon.

**Comment by rainbow1016** [comment]
very good question!

**Comment by JohannesHoppe** [comment]
This is a real showstopper now. We used a `404.html` file as a workaround for a long time, but now the Brave browser displays a warning if a 404 header is sent. Additionally, link previews aren't shown for pages that return a 404 header. As a consulting company, we stopped recommending Github Pages. Cloudflare Pages is a better alternative, that has real support for SPAs.
<img width="1495" alt="Screenshot 2024-01-22 at 19 34 29" src="https://github.com/community/community/assets/640639/e3cacbde-372a-47ce-9353-35f80e096ac8">

**Comment by IbrahimTanyalcin** [comment]
The SO post OP has linked offers a solution to the problem for people using react. 

I intutively realized using hashes would mimic the desired behavior to the extent possible. Using GitHub's public API and caching the contents returned from the API (without bearer tokens, to respect the limit) for a given ttl turned out to be a straight forward solution. This also allows one to extend the functionality using `.htaccess` like config files for each route (I used meta.json). 

I do not want this to be taken as spam but people who want to get around this issue using vanilla JS and nothing else, I have this repo here:

https://github.com/IbrahimTanyalcin/test-blog

Fork and start adding files/folders under static/md and it will work as expected. The root (static/md by default) can be reassigned to something else of couse.

**Reply by landsman** [reply]
It is spam. Already mentioned in the first post. Delete it.

**Reply by IbrahimTanyalcin** [reply]
What are you talking about?

**Comment by landsman** [comment]
It's really weird that GitHub Pages still do not support native rewrite.

**Comment by alemyaobed** [comment]
**SPA Routing on GitHub Pages: What Worked for Me**

I ran into the same issue with my Vite-based SPA — direct navigation to client-side routes like `/about` or `/dashboard` would return a 404 on GitHub Pages.

After trying the `--spa` flag with `gh-pages` (which doesn't exist), I implemented the simplest and most reliable workaround:

> 🔧 **Copy `index.html` to `404.html` in the build output folder before deploying.**

For example, in a GitHub Actions workflow:

```yaml
- run: cp dist/index.html dist/404.html
```
or just find a way to copy the index.html to the 404.html after the build but ensure that the router you are using or your routes handles 404 properly internally.

This works because GitHub Pages serves `404.html` for unknown routes, so your SPA boots up as expected, and the client-side router takes over.


Hope this helps someone else avoid the same rabbit hole.
If GitHub ever supports native rewrites for SPAs, that'll be a huge win 🚀

**Reply by GabenGar** [reply]
This will nuke your SEO.

**Comment by mrkrstphr** [comment]
This is a complete hack -- but I "solved" this by copying my `index.html` to every possible path, which ended up being dozens of files so far.

So if I have routes:

```
/
  about/
  products/big-television
  products/laptops
  products/hot-dog-buns
  contact-us/
```

I've generated:

```
/index.html <--- already existed
  about/index.html
  products/big-television/index.html
  products/laptops/index.html
  products/hot-dog-buns/index.html
  contact-us/index.html
```

This is pretty insane, but it fixes directly linking to pages and refreshing on a page, and doesn't suffer from the 404 problem.

I've scripted this as part of my build process so I don't have to worry about generating these. It works okay. Really hoping GitHub can add official support someday.
